src/CSharpTest.Net.Collections/IO/TransactedCompoundFile.cs

```diff
@@ -887,44 +898,89 @@ public Stream Read(ref BlockRef block, bool headerOnly, FGet fget)
                 do
                 {
                     retry = false;
-                    long position = _sectionPosition + (BlockSize*block.Offset);
-                    bytes = new byte[headerOnly ? BlockHeaderSize : block.ActualBlocks * BlockSize];
+                    long position = _sectionPosition + (BlockSize * block.Offset);
+                    var byteArrayLength = headerOnly ? BlockHeaderSize : block.ActualBlocks * BlockSize;
+
+                    bytes = _bytePool.Rent(byteArrayLength);
 
-                    readBytes = fget(position, bytes, 0, bytes.Length);
+                    readBytes = fget(position, bytes, 0, byteArrayLength);
                     if (readBytes < BlockHeaderSize)
-                        throw new InvalidDataException();
+                    {
+                        _bytePool.Return(bytes, true);
+                        throw new InvalidDataException("1");
+                    }
 
                     headerSize = bytes[OffsetOfHeaderSize];
-                    length = (int) GetUInt32(bytes, OffsetOfLength);
+                    length = (int)GetUInt32(bytes, OffsetOfLength);
 
-                    block.ActualBlocks = (int) GetUInt32(bytes, OffsetOfBlockCount);
+                    block.ActualBlocks = (int)GetUInt32(bytes, OffsetOfBlockCount);
                     uint blockId = GetUInt32(bytes, OffsetOfBlockId);
 
+                    if(headerSize < BlockHeaderSize)
+                    {
+                        throw new InvalidDataException("2A");
+                    }
+                    if (blockId != block.Identity)
+                    {
+                        throw new InvalidDataException("2B");
+                    }
+                    if (block.Count < 16 && block.ActualBlocks != block.Count)
+                    {
+                        throw new InvalidDataException("2C");
+                    }
+                    if ((block.Count == 16 && block.ActualBlocks < 16))
+                    {
+                        throw new InvalidDataException("2D");

```

Comment:
These messages looks like test data?

Reply:
Those were for debugging purposes. I've removed them so it's equivalent.